Mesh2faust eigen

khiner opened the pull request:
cc @sletz @rmichon @josmithiii

## Goal

This PR is about making `mesh2faust` work on non-Intel processors (such as M1 Macs) by removing the dependency on MKL. With this PR, I have successfully created `.lib` files for Romain's example bell `.obj` files [linked here](https://ccrma.stanford.edu/~rmichon/pmFaust/) on Ubuntu 20.10/x86_64 and MacOS Ventura/M1 Pro.

## Details

The only thing that MKL is being used for solving the sparse eigenvalue problem. After looking at the landscape I decided to go with Eigen and [Spectra](https://github.com/yixuan/spectra) instead of Pardiso/MKL for finding the eigenvalues/eigenvectors. This seems like a reasonable approach since the library is robust, focused on only this problem, header-only, and its only dependency is Eigen. I added Spectra as a submodule and directed users to install the Eigen library on their system using a package manager in the readme.

This PR removes all BLAS/LAPACK/Pardiso/MKL setup & dependency. Eigen doesn't depend on any of these, using its algorithms if none are linked. However, it does automatically defer to BLAS/LAPACK/MKL in some circumstances if one is linked (see [here](https://eigen.tuxfamily.org/dox/TopicUsingBlasLapack.html) for details). I also removed Vega's solver lib and all solver headers (except `CGSolver.h`, which is used in an example file).

## Runtime

Here is a gist showing a runtime comparison, along with a 1:1 comparison of the old vs new eigenvalues:
https://gist.github.com/khiner/c97788198bbc69b04d718128acdfcb34

I found a lot of variance in the runtimes, but this is pretty representative - it basically tends to take around the same amount of time.

## Not great things

One thing I don't like about this PR is that, since I haven't replaced the FEM parts of the code, it still relies on Vega and uses Vega's `SparseMatrix` classes in addition to the new `Eigen::SparseMatrix` class. This is just a minor annoyance, though, since it trades a bit of explicit conversion for getting rid of a whole bunch of more hidden baggage with the Vega/Pardiso/MKL chain.

Other annoying things are Eigen's default column-major ordering, and Spectra's largest-first ordering, which is different than Pardiso/Vega's smallest-first ordering. (This is more sensible, but it means a bit more index juggling.)

Thank you for taking a look! I'm happy to implement suggested changes etc., to get this to a point everyone's happy with.

khiner commented:
Update: Replaced Make with CMake build. Description updated.

khiner commented:
Update: Build as a static lib and as an executable.

khiner commented:
Update: Instead of allocating a mode gains matrix of size `nExPos x targetNModes`, only allocate a single `targetNModes` array.

    This was causing bad memory access when running on a worker thread with a small stack size.

khiner commented:
Update: Rebased to resolve `.gitmodules` conflict

khiner commented:
@sletz @rmichon @josmithiii Giving this a little bump. I know there's a lot here, and it's not a high priority, and I'm in no rush, but I think it would be nice to have these improvements in master-dev at some point since it resolves some pain points for using mesh2faust as a library in an application. I haven't provided much of any quantitative comparisons other than showing the eigenvalues converge to the same values, but this [demo video](https://www.youtube.com/watch?v=RwxgOHVBDvc) might serve as at least a qualitative sanity check that the changes here don't break the model behavior. If we want more confidence in model correctness, please let me know if you have ideas of what could help. I'm not sure how much this matters to you, but the code would still reflect all the details in the [original paper](https://hal.science/hal-03162901/document), even including mentions of Vega, except the mention of it being used as the eigenvalue solver.

For a quick reference, here is the kind of workflow I've used to test (again, qualitatively):
```shell
mkdir build && cd build && cmake .. && make
sudo cp mesh2faust /usr/local/bin
mesh2faust --infile englishBell/english.obj --nsynthmodes 5 --nfemmodes 10 --maxmode 10000 --expos 7840 5752 7682 7827 7823 --debug --material 1.06E11 0.34 8600
cd englishBell/
faust2cagtk englishBell.dsp
```

Thanks again for taking a look! Happy to make changes or add anything you think could help review.

josmithiii commented:
@khiner - Nice demo video!   I am fine with this PR myself, and I have forwarded the email to Sara Martin in case she is not tracking GitHub, but we need to hear from Romain Michon.  I also see that Stéphane is the last editor on many lines of the code.

Should the AppVeyor build failure be addressed?  I am not familiar with it.

Thanks for your useful contribution in this area!

sletz commented:
- I let @rmichon answer the general question of moving in this new direction
- I can help finishing the integration if/when needed

sletz commented:
> @khiner - Nice demo video! I am fine with this PR myself, and I have forwarded the email to Sara Martin in case she is not tracking GitHub, but we need to hear from Romain Michon. I also see that Stéphane is the last editor on many lines of the code.
> 
> Should the AppVeyor build failure be addressed? I am not familiar with it.
> 
> Thanks for your useful contribution in this area!
> 
> Incidentally, I tried to build mesh2audio on my M1 MacBook Pro (looks very interesting!), and I ran into this problem using the latest Faust master-dev:
> 
> ```
> CMake Error at lib/faust/build/misc/llvm.cmake:91 (message):
>   Cannot find program llvm-config (llvm-config or derived expected)
> Call Stack (most recent call first):
>   lib/faust/build/CMakeLists.txt:84 (llvm_config)
> ```
> 
> Does this ring a bell?

Is this happening when building Faust itself ? Or `mesh2audio` itself ?

josmithiii commented:
That's the mesh2audio build.

khiner commented:
@josmithiii Thank you for letting Sara know. I meant to mention that I didn't @ her in this PR since I couldn't find her GH username.

For the `llvm-config` error, building Faust as a library with CMake does assume that `llvm-config` is somewhere in your path. I should have put this in my Mac install instructions - if you installed llvm with homebrew, this should resolve your issue:

```shell
$ brew link llvm --force
```

If you installed llvm in some other way you might need to set up an alias yourself or consult the docs for the best way. Here's how mine is set up (from the `link llvm --force`):

```
$ ls -la /opt/homebrew/bin/llvm-config
lrwxr-xr-x  1 khiner  admin  37 Jun 10 13:03 /opt/homebrew/bin/llvm-config -> ../Cellar/llvm/16.0.5/bin/llvm-config
```

As you get something back from `which llvm-config` you should be good! Lmk if you run into anything else.


AppVeyor failures: I'm also not familiar with these. I believe they were breaking for everybody at the time of making this PR, but they seem to be working now. I'll rebase on master-dev and see if that resolves itself 🤞

khiner commented:
Ok, good to know about AppVeyor and sounds like it's expected for now.

Fwiw, David if you see that ping about llvm-config - it was already a requirement before your recent changes, so nothing to worry about. I remember running into this awhile ago when first setting up Faust for llvm-dsp compilation (incidentally, based on your TouchDesigner integration 😄).

josmithiii commented:
Thanks for the llvm-config fix. To do `brew link llvm --force`, I first had to do `brew unlink libomp`.
Now I make it to 
```
CMake Error at lib/CMakeLists.txt:69 (add_subdirectory):
  add_subdirectory given source "CDT/CDT" which is not an existing directory.
```
Simply creating the directory `app/build/lib/CDT/CDT/` did not work.
Oddly, CDT is empty, though the recursive pull behaved normally.

khiner commented:
@josmithiii How about just `lib/CDT/CDT` - is that empty? (Not the one in `build/`.) If so, something definitely went wrong during the recursive clone, and maybe we should try again from scratch:

```shell
$ git clone --recurse-submodules git@github.com:GATech-CSE-6730-Spring-2023-Project/mesh2audio.git
```

To keep this thread from getting too noisy I opened a GH issue in mesh2audio to troubleshoot further: https://github.com/GATech-CSE-6730-Spring-2023-Project/mesh2audio/issues/2

Thanks for your patience!

khiner commented:
@sletz @josmithiii I tried contacting @rmichon via email but haven't heard back. I am continuing some work along these lines, and it would be nice not to have to keep managing/referencing back to this branch.  What do you think is the best path forward to getting this merged to master-dev? Perhaps one of you is in closer contact with Romain and can politely point him to review this PR when he gets a chance? 😬 Thank you for your help so far, and especially @josmithiii for testing by building my app over the summer!

josmithiii commented:
How about we propose that you become the maintainer of mesh2faust instead of him?
Romain is spread very thin and will likely welcome that.
Thanks for your excellent work in this area.

sletz commented:
That make sense yes ! 
I can possibly help merge the PR. @rmichon what do you think ?

khiner commented:
> How about we propose that you become the maintainer of mesh2faust instead of him?
I'd be more than happy to!

> Romain is spread very thin and will likely welcome that.
I completely understand.

> Thanks for your excellent work in this area.
Thank you! I wouldn't have known how to begin without your (and Romain's) work to build on, not to mention all your books and educational materials.

rmichon commented:
Hi everyone,
Sorry about this very very late reply. I've been busy with many things recently and I must admit that I missed some of the github notifications. Thanks so much Karl for all these contributions which are of course totally welcome. Feel free to make any modifications to mesh2faust without asking me.
Cheers,
Romain

sletz commented:
OK, then @khiner I can help making the merge. Possibly in squash mode if we don't need to keep all all the comings and goings?  What do you think ?

khiner commented:
Thanks, for getting back on this, Romain, and no worries about the delay. Thanks for your fantastic work and documentation on this project and for being open to contributions!

@sletz I'd prefer to keep the commits if you're okay with having 19 commits in the history. I just rebased and added a `[mesh2faust]` prefix to all commits, and they're rebased on the master-dev branch, so all commits should be able to be rebased (or merged) directly onto master-dev. If you'd like to have these more condensed, I can combine some of the commits before merging as well.

orlarey commented:
Very nice work, Karl ! Thank you

sletz commented:
OK @khiner I'll have a look ASAP.

khiner commented:
Thanks @sletz !!
